### PR TITLE
Introduce component mapping and dependencies to rebuild dependent packages

### PR DESCRIPTION
**Pull request by xen0l**
This PR removes old depend.mk functionality, which was never really used and replaces it with a new one. The system then will allow things like bumping COMPONENT_REVISION of all dependents or in the future bumping COMPONENT_VERSION (getting data from https://repology.org). 

To use this functionality, one has to run `gmake rebuild-dependants REBUILD_FMRI=<package>`, which will rebuild all the packages that are depending on <package>. Please, keep in mind that it will modify only the Makefile, it will not git add, git commit or open a PR (this could be future feature requests IMO).

We tested the functionality with @alarcher on slightly beefier machine, where certain operations (generation of `mapping.json` and `dependencies.json` took only 2 minutes). On single core machine, it will take around 10-20 minutes. However, by default, we will use all the CPUs possible to speed up the task. These files contain information about the package dependencies (information is taken from REQUIRED_PACKAGES). 

Any feedback, reviews, questions are welcome.

Thanks @alarcher for hosting me, helping me and working on this together!

### CODE REVIEW

File: make-rules/depend.mk [hunk 1]
```diff
@@ -1,32 +1,19 @@
 #
-# CDDL HEADER START
+# This file and its contents are supplied under the terms of the
```
Comment (xen0l) [inline make-rules/depend.mk:2]: I deleted the file and recreated it from scratch. That's why it is shown like this.

**Comment by pyhalov** [comment]
depend.mk existed for a reason. We need a way to rebuild component A only after component B is built.

**Comment by normjacobs** [comment]
Yes, Back in the day when dinosaurs roamed the earth, depend.mk was 
invented to manage dependencies in top level builds.  We also added 
explicit dependencies into component Makefiles to deal with things like 
tcl/tk/expect where one reaches into the build of the other for data 
(yes it is naughty).  If the new dependency graph generation bits can 
spit out a set of dependencies for the top level make, like was being 
done in components/Makefile:43-55, It may very well be a suitable 
replacement.  I haven't looked at the changes in any detail.

     -Norm

**Comment by alarcher** [comment]
@xen0l Can we merge component-translate already in a separate PR?

**Comment by alarcher** [comment]
@normjacobs Thanks for the information, the idea was to generate the dependency graph based on the Makefiles (e.g. REQUIRED_PACKAGES) then be able to adapt the workflow based on such dependencies. This would simplify rebuilds after compatibility breakage.

**Comment by xen0l** [comment]
@alarcher @pyhalov I think we can merge this as it is. depend.mk should be recreated, but given it never was used I don't treat it with the utmost importance. There are more urgent things e.g. fixing `pkgdepend`.

**Comment by xen0l** [comment]
Also, for the record, the goal is it to replace `depends.mk` with proper build order in `components.mk`. I had this in mind when working on this.

**Comment by normjacobs** [comment]
> Also, for the record, the goal is it to replace `depends.mk` with proper build order in `components.mk`. I had this in mind when working on this.

FWIW, Without the replacement, It seems like removal of the depends.mk mechanism can have a negative impact on a top level build.

File: components/Makefile [hunk 2]
```diff
@@ -40,25 +40,24 @@ $(WS_TOP)/components/$(ENCUMBERED)components.mk:	# Makefile # $(WS_TOP)/componen
...
+# mapping.json is auto-generated by the build tools. It provides a map
 # between component path and component FMRI.
 $(WS_TOP)/components/mapping.json:
 	@echo "Generating component mapping..."
 	@$(WS_TOOLS)/userland-mapping --workspace=$(WS_TOP)
 mapping.json: $(WS_TOP)/components/mapping.json
 
+# dependencies.json is auto-generate by the build tools. It provides information
+# about component dependencies.
+$(WS_TOP)/components/dependencies.json: mapping.json
```
Comment (pyhalov) [inline components/Makefile:52]: Should we support separate dependencies.json for encumbered/*?

**Review by Mno-hime — APPROVED** [review]

File: components/Makefile (hunk 2 above)
Comment (xen0l) [reply components/Makefile:52]: I left this for simplicity reasons. What would be the benefit of that @pyhalov?

Comment (pyhalov) [reply components/Makefile:52]: The same as for other oi-userland dependencies - answering questions 'what can break if I touch component A'.

Comment (xen0l) [reply components/Makefile:52]: Well, @pyhalov, but it doesn't matter if you are using single file or multiple files, does it? :)

Comment (pyhalov) [reply components/Makefile:52]: it shouldn't

**Comment by xen0l** [comment]
@pyhalov / @alarcher could get one more approval, so i can merge this?

**Review by alarcher — APPROVED** [review]

**Comment by alarcher** [comment]
> @pyhalov / @alarcher could get one more approval, so i can merge this?

Modulo fixing conflicts

**Comment by Mno-hime** [comment]
Trying this for to-be-updated `image/library/librsvg`. Few "misses"?

```patch
--- a/components/editor/emacs/Makefile
+++ b/components/editor/emacs/Makefile
@@ -25,6 +25,7 @@ include ../../../make-rules/shared-macros.mk
 
 COMPONENT_NAME=                emacs
 COMPONENT_VERSION=     26.1
+COMPONENT_REVISION=    5
 COMPONENT_PROJECT_URL= https://www.gnu.org/software/emacs/
 COMPONENT_SRC=         $(COMPONENT_NAME)-$(COMPONENT_VERSION)
 COMPONENT_ARCHIVE=     $(COMPONENT_SRC).tar.xz
```
```patch
--- a/components/image/graphviz/Makefile
+++ b/components/image/graphviz/Makefile
@@ -27,7 +27,7 @@ include ../../../make-rules/shared-macros.mk
 
 COMPONENT_NAME=                graphviz
 COMPONENT_VERSION=     2.40.1
-COMPONENT_REVISION=    3
+COMPONENT_REVISION=    13
 COMPONENT_PROJECT_URL= http://www.graphviz.org/
 COMPONENT_SRC=         $(COMPONENT_NAME)-$(COMPONENT_VERSION)
 COMPONENT_ARCHIVE=     $(COMPONENT_SRC).tar.gz
```
```patch
--- a/components/library/gegl/Makefile
+++ b/components/library/gegl/Makefile
@@ -17,7 +17,7 @@ include ../../../make-rules/shared-macros.mk
 
 COMPONENT_NAME= gegl
 COMPONENT_VERSION= 0.2.0
-COMPONENT_REVISION= 1
+COMPONENT_REVISION=    3
 COMPONENT_SUMMARY= GEGL (Generic Graphics Library) is a graph based image processing framework
 COMPONENT_SRC= $(COMPONENT_NAME)-$(COMPONENT_VERSION)
 COMPONENT_ARCHIVE= $(COMPONENT_SRC).tar.bz2
```

**Comment by Mno-hime** [comment]
`$ gmake rebuild-dependants REBUILD_FMRI=library/glib2`:
```
Bumping COMPONENT_REVISION for dependants of library/glib2
...
Processing desktop/administration/nwam-manager
Traceback (most recent call last):
  File "/export/home/newman/ws/oi-userland/tools/userland-bump", line 139, in <module>
    main()
  File "/export/home/newman/ws/oi-userland/tools/userland-bump", line 135, in main
    rebuild_dependent_fmris(fmri=fmri, db_path=db_path, workspace=workspace, subdir=subdir, verbose=verbose)
  File "/export/home/newman/ws/oi-userland/tools/userland-bump", line 106, in rebuild_dependent_fmris
    contents = bump_component_revision(contents)
  File "/export/home/newman/ws/oi-userland/tools/userland-bump", line 79, in bump_component_revision
    component_revision = int(contents[component_revision_idx].split('=')[-1])
ValueError: invalid literal for int() with base 10: '$(shell cd $(COMPONENT_SRC); git rev-list HEAD --count)\n'
```
Line in question: `COMPONENT_REVISION=$(shell cd $(COMPONENT_SRC); git rev-list HEAD --count)`

**Comment by Mno-hime** [comment]
Here's the patch for the `ValueError`:
```patch
diff --git a/tools/userland-bump b/tools/userland-bump
index 82e8a3c38..a92a495f0 100755
--- a/tools/userland-bump
+++ b/tools/userland-bump
@@ -76,9 +76,15 @@ def bump_component_revision(contents):
     if has_component_revision:
         contents_new.extend(contents[:component_revision_idx])
 
-        component_revision = int(contents[component_revision_idx].split('=')[-1])
-        component_revision += 1
-        contents_new.append('COMPONENT_REVISION=\t{}\n'.format(component_revision))
+        component_revision = contents[component_revision_idx].split('=')[-1]
+        try:
+            component_revision_int = int(component_revision)
+        except ValueError:
+            print('\tSkipping component, COMPONENT_REVISION field malformed: {}'.format(component_revision))
+            return contents
+        else:
+            component_revision_int += 1
+            contents_new.append('COMPONENT_REVISION=\t{}\n'.format(component_revision_int))
 
         contents_new.extend(contents[component_revision_idx + 1:])
     else:
```

**Comment by pyhalov** [comment]
Can we reuse gmake print-COMPONENT_REVISION ?

**Comment by xen0l** [comment]
@pyhalov calling external commands is expensive. As for fixes, I don't have much time these days, so I might be able to fix these in foreseeable future.

**Comment by Mno-hime** [comment]
We really need this for OpenSSL 1.1.1.

**Comment by Mno-hime** [comment]
Rebased to fix conflicts.

**Comment by Mno-hime** [comment]
I am afraid it stuck for me, on a machine where it never run before, keeps "running" for 12 hours on 12 core VM:
```
  oi  newman   openssl  …  ws  oi-userland  components  time gmake rebuild-dependants REBUILD_FMRI=library/security/openssl
Generating component list...
Generating component mapping...
```
Running `truss` on `make` and `userland-mapping` processes shows that they are sleeping in one way or another.

**Comment by Mno-hime** [comment]
Got to cancel it:
```
 newman   openssl  ~  ws  oi-userland  components  time gmake rebuild-dependants REBUILD_FMRI=library/security/openssl
Generating component mapping...
^CTraceback (most recent call last):
  File "/export/home/newman/ws/oi-userland/tools/userland-mapping", line 109, in <module>
Process ForkPoolWorker-2:
Process ForkPoolWorker-4:
Process ForkPoolWorker-3:
Process ForkPoolWorker-1:
    main()
  File "/export/home/newman/ws/oi-userland/tools/userland-mapping", line 105, in main
    generate_userland_mapping(workspace_path=workspace, subdir=subdir)
  File "/export/home/newman/ws/oi-userland/tools/userland-mapping", line 81, in generate_userland_mapping
    results = pool.map(generate_component_data, paths)
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 266, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 638, in get
    self.wait(timeout)
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 635, in wait
Traceback (most recent call last):
    self._event.wait(timeout)
  File "/usr/lib/python3.5/threading.py", line 549, in wait
  File "/usr/lib/python3.5/multiprocessing/process.py", line 252, in _bootstrap
    self.run()
  File "/usr/lib/python3.5/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 44, in mapstar
    return list(map(*args))
  File "/export/home/newman/ws/oi-userland/tools/userland-mapping", line 66, in generate_component_data
    for out in proc.stdout:
KeyboardInterrupt
Traceback (most recent call last):
  File "/usr/lib/python3.5/multiprocessing/process.py", line 252, in _bootstrap
    self.run()
Traceback (most recent call last):
  File "/usr/lib/python3.5/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib/python3.5/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib/python3.5/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
  File "/usr/lib/python3.5/multiprocessing/process.py", line 252, in _bootstrap
    self.run()
  File "/usr/lib/python3.5/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib/python3.5/multiprocessing/queues.py", line 334, in get
    with self._rlock:
  File "/usr/lib/python3.5/multiprocessing/synchronize.py", line 96, in __enter__
    return self._semlock.__enter__()
KeyboardInterrupt
KeyboardInterrupt
    signaled = self._cond.wait(timeout)
  File "/usr/lib/python3.5/threading.py", line 293, in wait
    waiter.acquire()
KeyboardInterrupt
Traceback (most recent call last):
  File "/usr/lib/python3.5/multiprocessing/process.py", line 252, in _bootstrap
    self.run()
  File "/usr/lib/python3.5/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.5/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/usr/lib/python3.5/multiprocessing/queues.py", line 335, in get
    res = self._reader.recv_bytes()
  File "/usr/lib/python3.5/multiprocessing/connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
  File "/usr/lib/python3.5/multiprocessing/connection.py", line 407, in _recv_bytes
    buf = self._recv(4)
  File "/usr/lib/python3.5/multiprocessing/connection.py", line 379, in _recv
    chunk = read(handle, remaining)
KeyboardInterrupt
gmake: *** [Makefile:47: /export/home/newman/ws/oi-userland/components/mapping.json] Error 1

real	2155m46.930s
user	18m19.459s
sys	8m55.457s
```

**Comment by alarcher** [comment]
> I am afraid it stuck for me, on a machine where it never run before, keeps "running" for 12 hours on 12 core VM:
> 
> ```
>   oi  newman   openssl  …  ws  oi-userland  components  time gmake rebuild-dependants REBUILD_FMRI=library/security/openssl
> Generating component list...
> Generating component mapping...
> ```
> 
> Running `truss` on `make` and `userland-mapping` processes shows that they are sleeping in one way or another.

Found it.

**Pull request merged**